latte/src/main/java/gg/beemo/latte/util/SuspendingRatelimit.kt

```diff
@@ -6,32 +6,55 @@ import kotlinx.coroutines.sync.withPermit
 import kotlin.time.Duration
 
 class SuspendingRatelimit(private val burst: Int, private val duration: Duration) {
-
     @Volatile
     private var remainingQuota: Int = burst
+
     @Volatile
     private var resetTimestamp: Long = 0
 
     private val quotaRequestSem = Semaphore(1)
 
-    fun overrideRatelimit(remainingQuota: Int, resetTimestamp: Long) {
+    fun overrideRatelimit(
+        remainingQuota: Int,
+        resetTimestamp: Long,
+    ) {
         this.remainingQuota = remainingQuota
         this.resetTimestamp = resetTimestamp
     }
 
+    private fun calculateWaitTime(): Long {
+        return (resetTimestamp - System.currentTimeMillis()).coerceAtLeast(0)
+    }
+
+    private fun tryResetQuota() {
+        if (System.currentTimeMillis() >= resetTimestamp) {
+            remainingQuota = burst
+            resetTimestamp = System.currentTimeMillis() + duration.inWholeMilliseconds
+        }
+    }
+
     suspend fun requestQuota() {
         quotaRequestSem.withPermit {
             if (remainingQuota <= 0) {
-                val waitTime = (resetTimestamp - System.currentTimeMillis()).coerceAtLeast(0)
+                val waitTime = calculateWaitTime()
                 delay(waitTime)
             }
-            if (System.currentTimeMillis() >= resetTimestamp) {
-                remainingQuota = burst
-                resetTimestamp = System.currentTimeMillis() + duration.inWholeMilliseconds
-            }
+            tryResetQuota()
+
             check(remainingQuota > 0)
             remainingQuota--
         }
     }
 
+    fun tryRequestQuota(): Pair<Boolean, Long?>  {
+        if (remainingQuota <= 0) {
+            val waitTime = calculateWaitTime()
+            return false to waitTime
+        }
+        tryResetQuota()
+
+        check(remainingQuota > 0)
+        remainingQuota--
```

Comment:
Given thas these two lines are the same in the new and the old method, I think we should also move them to the common function and perhaps rename it to something a bit more generic then, like `calculateQuota`. Otherwise LGTM.

Reply:
Somehow, moving these two lines to a common function would result in the function misbehaving (failing), so yeah, I think we can just keep it like that, given that it's just two lines. What do you think? (I think it's because of the `check` that's failing).